### PR TITLE
Issue 79 implement data source insert convenience methods

### PR DESCRIPTION
**Pull request by psartzetakis**
## Pull request checklist

- [x] All tests pass. 
- [x] Demo project builds and runs.
- [x] I have resolved merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines) and [contributing guidelines](https://github.com/jessesquires/HowToContribute).

## What's in this pull request?
This pull request adds some` insert/append` methods to the `DataSouce`

### CODE REVIEW

File: Source/DataSource.swift [hunk 1]
```diff
@@ -162,6 +162,40 @@ public struct DataSource<S: SectionInfoProtocol>: DataSourceProtocol {
...
+    /**
+     Adds the item at the specified row and section.
+     
+     - parameter item:    The item to be inserted.
+     - parameter row:     The specified row location for the item.
+     - parameter section: The specified section location for the item.
+     */
+    public mutating func insert(item: S.Item, atRow row: Int, inSection section: Int) {
+        guard section < numberOfSections() else { return }
+        guard row <= numberOfItems(inSection: section) else { return }
```
Comment (jessesquires) [inline Source/DataSource.swift:185]: should be `row <`, right ?

File: Source/DataSource.swift [hunk 2]
```diff
@@ -162,6 +162,40 @@ public struct DataSource<S: SectionInfoProtocol>: DataSourceProtocol {
...
+    /**
+     Appends the item at the specified section.
+     
+     - parameter item:    The item to be appended.
+     - parameter section: The specified section location for the item.
+     */
+    public mutating func append(_ item: S.Item, inSection section: Int) {
+        guard let items = items(inSection: section) else { return }
+        insert(item: item, atRow: items.endIndex, inSection: section)
+    }
```
Comment (jessesquires) [inline Source/DataSource.swift:198]: 👍

File: Tests/DataSourceTests.swift [hunk 3]
```diff
@@ -241,6 +241,85 @@ final class DataSourceTests: XCTestCase {
...
+        let newItem = FakeViewModel()
+        let ip = IndexPath(item: 2, section: 1)
+        dataSource.insert(item: newItem, at: ip)
+        
+        let sectionItems = dataSource.items(inSection: ip.section)
+        let insertedIndex = sectionItems?.index(of: newItem)
+        
+        //THEN: the item is inserted at the specific index path
+        XCTAssertNotNil(sectionItems)
+        XCTAssertNotNil(insertedIndex)
```
Comment (jessesquires) [inline Tests/DataSourceTests.swift]: let's remove these 2 nil checks

File: Tests/DataSourceTests.swift [hunk 4]
```diff
@@ -241,6 +241,85 @@ final class DataSourceTests: XCTestCase {
...
+        XCTAssertEqual(insertedIndex, ip.row, "New item should be at the requested index")
```
Comment (jessesquires) [inline Tests/DataSourceTests.swift:259]: let's also test `Equal(dataSource[ip], newItem)`

File: Tests/DataSourceTests.swift [hunk 5]
```diff
@@ -241,6 +241,85 @@ final class DataSourceTests: XCTestCase {
...
+        let section = 1
+        dataSource.append(newItem, inSection: section)
+        
+        let sectionItems = dataSource.items(inSection: section)
+        let insertedIndex = sectionItems?.index(of: newItem)
+        
+        //THEN: the item is appended in the specific section
+        XCTAssertNotNil(sectionItems)
+        XCTAssertNotNil(insertedIndex)
+        XCTAssertEqual(insertedIndex, sectionItems!.count - 1, "New item should be at the requested index")
```
Comment (jessesquires) [inline Tests/DataSourceTests.swift:278]: same 2 comments as above

File: Tests/DataSourceTests.swift [hunk 6]
```diff
@@ -241,6 +241,85 @@ final class DataSourceTests: XCTestCase {
...
+    }
+
+    func test_thatDataSource_doesntInsertData_atOutOfBounds_Section() {
```
Comment (jessesquires) [inline Tests/DataSourceTests.swift]: nit: `..._doesNotInsert...`

File: Tests/DataSourceTests.swift [hunk 7]
```diff
@@ -241,6 +241,85 @@ final class DataSourceTests: XCTestCase {
...
+        let sectionB = Section(items: FakeViewModel(), FakeViewModel(), footerTitle: "Footer")
+        var dataSource = DataSource(sections: sectionA, sectionB)
+        
+        // WHEN: we add an item at an index that doesn't exist
+        let newItem = FakeViewModel()
+        let ip = IndexPath(item: 4, section: 1)
+        dataSource.insert(item: newItem, at: ip)
+        
+        let sectionItems = dataSource.items(inSection: 1)
+        let contains = sectionItems!.contains(newItem)
```
Comment (jessesquires) [inline Tests/DataSourceTests.swift]: let's just put this inline in the `assertFalse`

also, we use `sectionItems?` instead?

**Review by jessesquires — CHANGES_REQUESTED** [review]
Thanks @psartzetakis !

Just a few minor comments 💯

File: Source/DataSource.swift (hunk 1 above)
Comment (psartzetakis) [reply Source/DataSource.swift:185]: Well if he have 2 rows (`row-0`, `row-1`) and the user wants an item at `row-2` then the `row` will be `2` and the `numberOfItems(inSection: section)` will be already `2`. So the `equality` is for `appending`, correct? Do I miss something..?🤔

Comment (jessesquires) [reply Source/DataSource.swift:185]: oops, you're right. 😅 😊

**Review by jessesquires — APPROVED** [review]
Awesome! Thanks @psartzetakis ! 💯 🥇

**Pull request merged**